rpc/src/client/event_listener.rs

```diff
@@ -83,33 +88,81 @@ impl EventListener {
         Ok(())
     }
 
-    /// Get the next event from the websocket
-    pub async fn get_event(&mut self) -> Result<Option<ResultEvent>, RPCError> {
-        let msg = self
-            .socket
-            .next()
-            .await
-            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+    /// Get the next event from the websocket. Automatically handles websocket
+    /// protocol details, like responding to ping messages, so it can either
+    /// produce events or errors.
+    pub async fn get_event(&mut self) -> Result<ResultEvent, RPCError> {
+        loop {
+            let msg = self
+                .socket
+                .next()
+                .await
+                .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
 
-        if let Ok(result_event) = serde_json::from_str::<WrappedResultEvent>(&msg.to_string()) {
-            // if we get an rpc error here, we will bubble it up:
-            return Ok(Some(result_event.into_result()?));
+            match msg {
+                Message::Text(msg_text) => {
+                    match serde_json::from_str::<WrappedResultEvent>(msg_text.as_str()) {
+                        // if we get an rpc error here, we will bubble it up:
+                        Ok(result_event) => return Ok(result_event.into_result()?),
+                        Err(e) => {
+                            return Err(RPCError::new(
+                                Code::Other(-1),
+                                Some(format!(
+                                    "failed to decode incoming message as an event: {}",
+                                    e
+                                )),
+                            ))
+                        }
+                    }
+                }
+                Message::Ping(_) => {
+                    self.socket
+                        .send(Message::Pong(Vec::new()))
```

Comment:
Did you try that against a running tendermint node? Not sure any of these will sent in the integration test below 🤔

Reply:
I did run it against a Tendermint node, and the pings did come through (I used `dbg!` statements to show how we send pong responses, but subsequently removed the `dbg!` statements to clean things up a little).

We definitely need more extensive testing here, but I'm still debating exactly how to go about doing that. Will definitely be looking into more extensive testing in a future PR.